stores/slabbuffer.go

```diff
@@ -0,0 +1,505 @@
+	dir                             string
+	s                               *SQLStore
+
+	mu                sync.Mutex
+	completeBuffers   map[bufferGroupID][]*SlabBuffer
+	incompleteBuffers map[bufferGroupID][]*SlabBuffer
+	buffersByKey      map[string]*SlabBuffer
+}
+
+func newSlabBufferManager(sqlStore *SQLStore, slabBufferCompletionThreshold int64, partialSlabDir string) (*SlabBufferManager, error) {
+	if slabBufferCompletionThreshold < 0 || slabBufferCompletionThreshold > 1<<22 {
+		return nil, fmt.Errorf("invalid slabBufferCompletionThreshold %v", slabBufferCompletionThreshold)
+	}
+
+	// load existing buffers
+	var buffers []dbBufferedSlab
+	err := sqlStore.db.
+		Joins("DBSlab").
+		Find(&buffers).
+		Error
+	if err != nil {
+		return nil, err
+	}
+	mgr := &SlabBufferManager{
+		bufferedSlabCompletionThreshold: slabBufferCompletionThreshold,
+		dir:                             partialSlabDir,
+		s:                               sqlStore,
+		completeBuffers:                 make(map[bufferGroupID][]*SlabBuffer),
+		incompleteBuffers:               make(map[bufferGroupID][]*SlabBuffer),
+		buffersByKey:                    make(map[string]*SlabBuffer),
+	}
+	for _, buffer := range buffers {
+		if buffer.DBSlab.ID == 0 {
+			// Buffer doesn't have a slab. We can delete it.
+			if err := sqlStore.db.Delete(&buffer).Error; err != nil {
+				return nil, fmt.Errorf("failed to delete buffer %v: %v", buffer.ID, err)
+			}
+			if err := os.RemoveAll(filepath.Join(partialSlabDir, buffer.Filename)); err != nil {
+				return nil, fmt.Errorf("failed to remove buffer file %v: %v", buffer.Filename, err)
+			}
+			continue
+		}
+		// Open the file.
+		file, err := os.OpenFile(filepath.Join(partialSlabDir, buffer.Filename), os.O_RDWR, 0600)
+		if err != nil {
+			return nil, err
+		}
+		// Create the slab buffer.
+		var ec object.EncryptionKey
+		if err := ec.UnmarshalText(buffer.DBSlab.Key); err != nil {
+			return nil, err
+		}
+		sb := &SlabBuffer{
+			dbID:     buffer.ID,
+			filename: buffer.Filename,
+			slabKey:  ec,
+			maxSize:  int64(bufferedSlabSize(buffer.DBSlab.MinShards)),
+			file:     file,
+			dbSize:   buffer.Size,
+			size:     buffer.Size,
+		}
+		// Add the buffer to the manager.
+		gid := bufferGID(buffer.DBSlab.MinShards, buffer.DBSlab.TotalShards, uint32(buffer.DBSlab.DBContractSetID))
+		if buffer.Complete {
+			mgr.completeBuffers[gid] = append(mgr.completeBuffers[gid], sb)
+		} else {
+			mgr.incompleteBuffers[gid] = append(mgr.completeBuffers[gid], sb)
+		}
+		mgr.buffersByKey[sb.slabKey.String()] = sb
+	}
+	return mgr, nil
+}
+
+func bufferGID(minShards, totalShards uint8, contractSet uint32) bufferGroupID {
+	var bgid bufferGroupID
+	bgid[0] = minShards
+	bgid[1] = totalShards
+	binary.LittleEndian.PutUint32(bgid[2:], contractSet)
+	return bgid
+}
+
+func (mgr *SlabBufferManager) Close() error {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	for _, buffers := range mgr.buffersByKey {
+		if err := buffers.file.Close(); err != nil {
```

Comment:
`nit` should we collect the errors and try and close the other buffer files here